### PR TITLE
HDDS-5377. Add a Namespace Summary tab to Recon Web UI

### PR DESCRIPTION
**Pull request by yuangu002**
## What changes were proposed in this pull request?

Add a new tab to the Recon Web UI about Namespace Summary.

[Other fixes]
1. Fixed NPE bugs in `NSSummaryEndpoint` when bucket or directory is empty.
2. Added a isKey field in DU's subpath in order for the UI to differentiate directories from keys.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5377

## How was this patch tested?
Frontend mocking: `pnpm run dev`
Demo: https://www.youtube.com/watch?v=Q6naW-BxpuQ

On docker:
Configs:
```
ozone sh volume create /vol1
ozone sh volume create /vol2
ozone sh bucket create /vol1/bucket11
ozone sh bucket create /vol1/bucket12
ozone sh bucket create /vol2/bucket21
ozone sh bucket create /vol2/bucket22
ozone sh key put --replication=THREE /vol1/bucket11/key1 README.md
ozone sh key put --replication=THREE /vol1/bucket11/key2s README.md
ozone sh key put --replication=THREE /vol1/bucket11/dir1/key3 README.md
ozone sh key put /vol2/bucket21/key4 CONTRIBUTING.md
ozone sh key put /vol2/bucket21/key5 CONTRIBUTING.md
ozone sh key put /vol2/bucket22/key6 README.md
ozone sh key put --replication=THREE /vol2/bucket22/dir22/key7 LICENSE.txt
ozone sh key put /vol1/bucket12/dir1-1/key8 README.md
ozone sh key put /vol1/bucket12/dir1-2/key9 CONTRIBUTING.md
ozone sh key put --replication=THREE /vol1/bucket12/directKey README.md
```
Demo: https://www.youtube.com/watch?v=J7n-2kbNu-4

[Other]
I also use a shell script to quickly run docker-up with FSO enabled:
```
cd hadoop-ozone/dist/target/ozone-*-SNAPSHOT/compose/ozone
echo 'OZONE-SITE.XML_ozone.om.enable.filesystem.paths=true;' >> docker-config
echo 'OZONE-SITE.XML_ozone.om.metadata.layout=PREFIX;' >> docker-config
sed -i '' 's/-SIMPLE/-PREFIX/g' docker-compose.yaml
sed -i '' 's/-false/-true/g' docker-compose.yaml
docker-compose up -d --scale datanode=3
```

[Backlog]
1. When clicking a key, directly pops up the right drawer panel, which shows the metadata summary for the key
2. Add an auto-reloader to reload pie chart for the current input path in real-time

### CODE REVIEW

**Comment by vivekratnavel** [comment]
@yuangu002 Great job! Thanks for the demo video. Overall looks good to me. Just a couple of items to remember to improve the user experience:

- Rename "< Back" button to just "<" or "< Previous" to avoid confusion with browser back button behavior and make it disabled at the first visit when there is no history.
- Make the [display limit](https://github.com/apache/ozone/pull/2530/files#diff-4d9ddef7873bb3ee5f89fd7b50f68cc78eef723e37a764c16d6ef8c723ed2bcdR28) configurable in UI with a dropdown (5, 10, 15, 20, All). Then we should make the default be 10 since I think even 20 will be clunky.  

Both of these items can be tracked separately.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+import React from 'react';
+import axios from 'axios';
+import Plot from 'react-plotly.js';
+import {Row, Col, Icon, Button, Input} from 'antd';
+import {DetailPanel} from 'components/rightDrawer/rightDrawer';
+import * as Plotly from 'plotly.js';
+import {showDataFetchError} from 'utils/common';
+import './diskUsage.less';
+
+const DISPLAY_LIMIT = 20;
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:28]: Let's set this to `10` by default here?
```suggestion
const DISPLAY_LIMIT = 10;
```

**Comment by yuangu002** [comment]
[Fixed]
1. Add a dropdown list for configure display limit. Default value set to 10.
(Every time a pie section is clicked, go-back button is clicked, or a new path entered, default display is reset)
2. Make pie chart's title one line and all fonts 14 to avoid text overlapping.
3. Add a refresh button to refresh the current path and the currently selected display limit.
4. Add mock path "/clunky" to test display limit in frontend.

**Comment by yuangu002** [comment]
@smengcl @vivekratnavel would you take another look?

**Review by smengcl — APPROVED** [review]

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        <div className='page-header'>
+          Disk Usage
+        </div>
+        <div className='content-div'>
+          {isLoading ? <span><Icon type='loading'/> Loading...</span> : (
+            <div>
+              <Row>
+                <Col>
+                  <div className='go-back-button'>
+                    <Button type='primary' onClick={e => this.goBack(e, returnPath)}><Icon type='left'/> </Button>
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:378]:
```suggestion
                    <Button type='primary' onClick={e => this.goBack(e, returnPath)}><Icon type='left'/></Button>
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+                  </div>
+                  <div className='input-bar'>
+                    <h3>Path</h3>
+                    <form className='input' id='input-form' onSubmit={this.handleSubmit}>
+                      <Input placeholder='/' value={inputPath} onChange={this.handleChange}/>
+                    </form>
+                  </div>
+                  <div className='go-back-button'>
+                    <Button type='primary' onClick={e => this.refreshCurPath(e, returnPath)}><Icon type='redo'/> </Button>
```
Comment (smengcl) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:387]:
```suggestion
                    <Button type='primary' onClick={e => this.refreshCurPath(e, returnPath)}><Icon type='redo'/></Button>
```

**Review by smengcl — COMMENTED** [review]
The extra _space_ is causing the label icon to be not aligned to the center.

Removing them fixes the alignment issue as I tested.

image

**Comment by smengcl** [comment]
Merged. Thanks @yuangu002 for the contribution. Thanks @vivekratnavel for reviewing this.

**Pull request merged**